src/java.base/share/classes/java/lang/ref/Reference.java

```diff
@@ -235,61 +228,16 @@ private void enqueueFromPending() {
         if (q != ReferenceQueue.NULL) q.enqueue(this);
     }
 
-    private static final Object processPendingLock = new Object();
-    private static boolean processPendingActive = false;
-
     private static void processPendingReferences() {
         // Only the singleton reference processing thread calls
         // waitForReferencePendingList() and getAndClearReferencePendingList().
-        // These are separate operations to avoid a race with other threads
-        // that are calling waitForReferenceProcessing().
         waitForReferencePendingList();
-        Reference<?> pendingList;
-        synchronized (processPendingLock) {
-            pendingList = getAndClearReferencePendingList();
-            processPendingActive = true;
-        }
+        Reference<?> pendingList = getAndClearReferencePendingList();
         while (pendingList != null) {
             Reference<?> ref = pendingList;
             pendingList = ref.discovered;
             ref.discovered = null;
-
-            if (ref instanceof Cleaner) {
-                ((Cleaner)ref).clean();
-                // Notify any waiters that progress has been made.
-                // This improves latency for nio.Bits waiters, which
-                // are the only important ones.
-                synchronized (processPendingLock) {
-                    processPendingLock.notifyAll();
-                }
-            } else {
-                ref.enqueueFromPending();
-            }
-        }
-        // Notify any waiters of completion of current round.
-        synchronized (processPendingLock) {
-            processPendingActive = false;
-            processPendingLock.notifyAll();
-        }
-    }
-
-    // Wait for progress in reference processing.
-    //
-    // Returns true after waiting (for notification from the reference
-    // processing thread) if either (1) the VM has any pending
-    // references, or (2) the reference processing thread is
-    // processing references. Otherwise, returns false immediately.
-    private static boolean waitForReferenceProcessing()
-        throws InterruptedException
-    {
-        synchronized (processPendingLock) {
-            if (processPendingActive || hasReferencePendingList()) {
```

Comment:
It looks like `hasReferencePendingList()` is no longer used.

Reply:
Right, we can clean this up right here too. Done in new commit.